# JMAK Model to Find the Relationship Between Cellular Properties and DNA Repair

## Introduction
The genome, as the hereditary unit of life, is constantly under threat to be altered by genotoxic
agents and even the cellâ€™s own metabolism. Repair mechanisms are therefore essential to ensure
DNA integrity and to enable life itself. One of the best studied DNA repair mechanisms is the
nucleotide excision repair (NER) pathway, which is the only method that can remove UV-induced
cyclobutane pyrimidine dimers (CPDs) in higher eukaryotes. *In vitro* experiments revealed the
existence of two different lesion recognition mechanisms. The first one is coupled to transcription
(transcription-coupled repair, TCR) and is thus only available at genes. The other act genome-wide
(global-genome wide repair, GGR). After the initial recognition of the DNA break, the excision and
replacement behaviour is the same in both GGR and TCR. Despite the fact that these subpathways
are similar, they are usually related to different efficiency and speed. The scientific community,
however, has not reached yet a consensus what biological properties contribute to these different
dynamics.

This project aims to explain the removal of CPD lesion in yeast cells over time. It makes use of the
JMAK model which describes the phase shift of solids and the fraction of transformed particles by 
exploiting the expansion of patterns. These found parameters can be linked to real biological features.

## Requirements and Installation
The code requires `python3.8` if plotting is used. Otherwise any other python version >=3 should work.
It is recommended to have `pip` installed to deploy the necessary requirements. Once done, run

```commandline
python3.8 -m pip install -r requirements.txt
```

##